### ISSUE TITLE
[QUESTION] Receiving Video frame from remote

### ISSUE THREAD

**Issue by YUN-CHIEN-LO** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#779]
Hi, I want to show remote video ( send from viewer) and display them on local-end (master).

What I did is that I add a `g_signal_connect` in `receiveGstreamerAudioVideo`
```
    if (appsrcVideo != NULL) {
        g_signal_connect(appsrcVideo, "enough-data", G_CALLBACK(on_new_remote_video), (gpointer) pSampleStreamingSession);
    }
```
and then 
```
GstFlowReturn on_new_remote_video(GstElement* src, gpointer data)
{
    return on_new_remote(src, data, DEFAULT_VIDEO_TRACK_ID);
}
```
```
GstFlowReturn on_new_remote(GstElement* appsrc, gpointer data, UINT64 trackid){
...
// I plan to use gst_parse_launch to play video from appsrc
  return;
}
```

My problem is that I'm not sure if it really received video data, because if it did, I suppose the function `on_new_remote` will be repeatedly called, but instead it would be called only once.

Here is my modified `receiveGstreamerAudioVideo`
```
PVOID receiveGstreamerAudioVideo(PVOID args)
{
    STATUS retStatus = STATUS_SUCCESS;
    GstElement *pipeline = NULL, *appsrcAudio, *appsrcVideo = NULL;
    GstBus* bus;
    GstMessage* msg;
    GError* error = NULL;
    PSampleStreamingSession pSampleStreamingSession = (PSampleStreamingSession) args;
    gchar *videoDescription = "", *audioDescription = "", *audioVideoDescription;

    if (pSampleStreamingSession == NULL) {
        printf("[KVS GStreamer Master] receiveGstreamerAudioVideo(): operation returned status code: 0x%08x \n", STATUS_NULL_ARG);
        goto CleanUp;
    }

    // TODO: Wire video up with gstreamer pipeline

    switch (pSampleStreamingSession->pAudioRtcRtpTransceiver->receiver.track.codec) {
        case RTC_CODEC_OPUS:
            audioDescription = "appsrc name=appsrc-audio ! opusparse ! decodebin ! autoaudiosink";
            break;

        case RTC_CODEC_MULAW:
        case RTC_CODEC_ALAW:
            audioDescription = "appsrc name=appsrc-audio ! rawaudioparse ! decodebin ! autoaudiosink";
            break;
        default:
            break;
    }

    switch (pSampleStreamingSession->pVideoRtcRtpTransceiver->receiver.track.codec) {
        case RTC_CODEC_H264_PROFILE_42E01F_LEVEL_ASYMMETRY_ALLOWED_PACKETIZATION_MODE:
            videoDescription = "appsrc name=appsrc-video ! decodebin ! video/x-raw,width=1280,height=720,framerate=20/1  !  autovideosink";
            break;
        case RTC_CODEC_VP8:
            break;
        default:
            break;
    }

    audioVideoDescription = g_strjoin(" ", audioDescription, videoDescription, NULL);

    pipeline = gst_parse_launch(audioVideoDescription, &error);

    appsrcAudio = gst_bin_get_by_name(GST_BIN(pipeline), "appsrc-audio");
    if (appsrcAudio == NULL) {
        printf("[KVS GStreamer Master] gst_bin_get_by_name(): cant find audio appsrc, operation returned status code: 0x%08x \n", STATUS_INTERNAL_ERROR);
        goto CleanUp;
    }

    appsrcVideo = gst_bin_get_by_name(GST_BIN(pipeline), "appsrc-video");
    if (appsrcVideo == NULL) {
        printf("[KVS GStreamer Master] gst_bin_get_by_name(): cant find video appsrc, operation returned status code: 0x%08x \n", STATUS_INTERNAL_ERROR);
        goto CleanUp;
    }


    transceiverOnFrame(pSampleStreamingSession->pAudioRtcRtpTransceiver, (UINT64) appsrcAudio, onGstAudioFrameReady);
    transceiverOnFrame(pSampleStreamingSession->pVideoRtcRtpTransceiver, (UINT64) appsrcVideo, onGstVideoFrameReady);

    retStatus = streamingSessionOnShutdown(pSampleStreamingSession, (UINT64) appsrcAudio, onSampleStreamingSessionShutdown);
    if (retStatus != STATUS_SUCCESS) {
        printf("[KVS GStreamer Master] streamingSessionOnShutdown(): operation returned status code: 0x%08x \n", STATUS_INTERNAL_ERROR);
        goto CleanUp;
    }

    g_free(audioVideoDescription);

    if (pipeline == NULL) {
        printf("[KVS GStreamer Master] receiveGstreamerAudioVideo(): Failed to launch gstreamer, operation returned status code: 0x%08x \n",
               STATUS_INTERNAL_ERROR);
        goto CleanUp;
    }
    if (appsrcVideo != NULL) {
        g_signal_connect(appsrcVideo, "enough-data", G_CALLBACK(on_new_remote_video), (gpointer) pSampleStreamingSession);
    }else{}
    gst_element_set_state(pipeline, GST_STATE_PLAYING);

    /* block until error or EOS */
    bus = gst_element_get_bus(pipeline);
    msg = gst_bus_timed_pop_filtered(bus, GST_CLOCK_TIME_NONE, GST_MESSAGE_ERROR | GST_MESSAGE_EOS);

    /* Free resources */
    if (msg != NULL) {
        gst_message_unref(msg);
    }
    gst_object_unref(bus);
    gst_element_set_state(pipeline, GST_STATE_NULL);
    gst_object_unref(pipeline);

CleanUp:
    if (error != NULL) {
        printf("%s", error->message);
        g_clear_error(&error);
    }

    return (PVOID)(ULONG_PTR) retStatus;
}
```
I have done some searching online, and some said that the above code should be working, and by 'working' I'm hoping it means displaying received content.

Please give me some advice, or if there are any source code I can reference on.
Thank you.

**Comment by YUN-CHIEN-LO** [comment]
I have trace code to `PeerConnection.c` > `onFrameReadyFunc()`
I can see here frames from remote client is received. How can I display them?

**Comment by MushMal** [comment]
@YUN-CHIEN-LO seems that you are using GStreamer. Perhaps, the GStreamer forums might have more info on this. You need to decode the frames and render them. You might need to construct a pipeline for that and acts as a source for the frames.

**Comment by disa6302** [comment]
@YUN-CHIEN-LO , 

If you are using `DLOGx` to display the entry in the callback `on_new_remote_video`, make sure the appropriate log level is enabled. As such, the callback registration looks fine. I am going to close this issue for now since there is not much we can do for the GStreamer pipeline. Feel free to open a new issue if you have questions. Thanks!

**Comment by tonyong99** [comment]
> Hi, I want to show remote video ( send from viewer) and display them on local-end (master).
> 
> What I did is that I add a `g_signal_connect` in `receiveGstreamerAudioVideo`
> 
> ```
>     if (appsrcVideo != NULL) {
>         g_signal_connect(appsrcVideo, "enough-data", G_CALLBACK(on_new_remote_video), (gpointer) pSampleStreamingSession);
>     }
> ```
> 
> and then
> 
> ```
> GstFlowReturn on_new_remote_video(GstElement* src, gpointer data)
> {
>     return on_new_remote(src, data, DEFAULT_VIDEO_TRACK_ID);
> }
> ```
> 
> ```
> GstFlowReturn on_new_remote(GstElement* appsrc, gpointer data, UINT64 trackid){
> ...
> // I plan to use gst_parse_launch to play video from appsrc
>   return;
> }
> ```
> 
> My problem is that I'm not sure if it really received video data, because if it did, I suppose the function `on_new_remote` will be repeatedly called, but instead it would be called only once.
> 
> Here is my modified `receiveGstreamerAudioVideo`
> 
> ```
> PVOID receiveGstreamerAudioVideo(PVOID args)
> {
>     STATUS retStatus = STATUS_SUCCESS;
>     GstElement *pipeline = NULL, *appsrcAudio, *appsrcVideo = NULL;
>     GstBus* bus;
>     GstMessage* msg;
>     GError* error = NULL;
>     PSampleStreamingSession pSampleStreamingSession = (PSampleStreamingSession) args;
>     gchar *videoDescription = "", *audioDescription = "", *audioVideoDescription;
> 
>     if (pSampleStreamingSession == NULL) {
>         printf("[KVS GStreamer Master] receiveGstreamerAudioVideo(): operation returned status code: 0x%08x \n", STATUS_NULL_ARG);
>         goto CleanUp;
>     }
> 
>     // TODO: Wire video up with gstreamer pipeline
> 
>     switch (pSampleStreamingSession->pAudioRtcRtpTransceiver->receiver.track.codec) {
>         case RTC_CODEC_OPUS:
>             audioDescription = "appsrc name=appsrc-audio ! opusparse ! decodebin ! autoaudiosink";
>             break;
> 
>         case RTC_CODEC_MULAW:
>         case RTC_CODEC_ALAW:
>             audioDescription = "appsrc name=appsrc-audio ! rawaudioparse ! decodebin ! autoaudiosink";
>             break;
>         default:
>             break;
>     }
> 
>     switch (pSampleStreamingSession->pVideoRtcRtpTransceiver->receiver.track.codec) {
>         case RTC_CODEC_H264_PROFILE_42E01F_LEVEL_ASYMMETRY_ALLOWED_PACKETIZATION_MODE:
>             videoDescription = "appsrc name=appsrc-video ! decodebin ! video/x-raw,width=1280,height=720,framerate=20/1  !  autovideosink";
>             break;
>         case RTC_CODEC_VP8:
>             break;
>         default:
>             break;
>     }
> 
>     audioVideoDescription = g_strjoin(" ", audioDescription, videoDescription, NULL);
> 
>     pipeline = gst_parse_launch(audioVideoDescription, &error);
> 
>     appsrcAudio = gst_bin_get_by_name(GST_BIN(pipeline), "appsrc-audio");
>     if (appsrcAudio == NULL) {
>         printf("[KVS GStreamer Master] gst_bin_get_by_name(): cant find audio appsrc, operation returned status code: 0x%08x \n", STATUS_INTERNAL_ERROR);
>         goto CleanUp;
>     }
> 
>     appsrcVideo = gst_bin_get_by_name(GST_BIN(pipeline), "appsrc-video");
>     if (appsrcVideo == NULL) {
>         printf("[KVS GStreamer Master] gst_bin_get_by_name(): cant find video appsrc, operation returned status code: 0x%08x \n", STATUS_INTERNAL_ERROR);
>         goto CleanUp;
>     }
> 
> 
>     transceiverOnFrame(pSampleStreamingSession->pAudioRtcRtpTransceiver, (UINT64) appsrcAudio, onGstAudioFrameReady);
>     transceiverOnFrame(pSampleStreamingSession->pVideoRtcRtpTransceiver, (UINT64) appsrcVideo, onGstVideoFrameReady);
> 
>     retStatus = streamingSessionOnShutdown(pSampleStreamingSession, (UINT64) appsrcAudio, onSampleStreamingSessionShutdown);
>     if (retStatus != STATUS_SUCCESS) {
>         printf("[KVS GStreamer Master] streamingSessionOnShutdown(): operation returned status code: 0x%08x \n", STATUS_INTERNAL_ERROR);
>         goto CleanUp;
>     }
> 
>     g_free(audioVideoDescription);
> 
>     if (pipeline == NULL) {
>         printf("[KVS GStreamer Master] receiveGstreamerAudioVideo(): Failed to launch gstreamer, operation returned status code: 0x%08x \n",
>                STATUS_INTERNAL_ERROR);
>         goto CleanUp;
>     }
>     if (appsrcVideo != NULL) {
>         g_signal_connect(appsrcVideo, "enough-data", G_CALLBACK(on_new_remote_video), (gpointer) pSampleStreamingSession);
>     }else{}
>     gst_element_set_state(pipeline, GST_STATE_PLAYING);
> 
>     /* block until error or EOS */
>     bus = gst_element_get_bus(pipeline);
>     msg = gst_bus_timed_pop_filtered(bus, GST_CLOCK_TIME_NONE, GST_MESSAGE_ERROR | GST_MESSAGE_EOS);
> 
>     /* Free resources */
>     if (msg != NULL) {
>         gst_message_unref(msg);
>     }
>     gst_object_unref(bus);
>     gst_element_set_state(pipeline, GST_STATE_NULL);
>     gst_object_unref(pipeline);
> 
> CleanUp:
>     if (error != NULL) {
>         printf("%s", error->message);
>         g_clear_error(&error);
>     }
> 
>     return (PVOID)(ULONG_PTR) retStatus;
> }
> ```
> 
> I have done some searching online, and some said that the above code should be working, and by 'working' I'm hoping it means displaying received content.
> 
> Please give me some advice, or if there are any source code I can reference on.
> Thank you.

Hi, I am also trying to do the same but unable to find resolution to this post, even searched GStreamer forum.  thanks

**Comment by orchydea** [comment]
@JustRun-dotcom, did you manage to receive the video from the viewer? I'm facing the same problem.
@disa6302, @MushMal can you or the team point us in the right direction about how to properly configure the gstreamer video receiver pipeline? The audio receiver pipeline in the gstreamer master sample works out of the box, it would be great if you could provide some sample code about how to make the video receiver work with the [kinesis webrtc test page](https://awslabs.github.io/amazon-kinesis-video-streams-webrtc-sdk-js/examples/index.html)

**Comment by roggerfq** [comment]
@YUN-CHIEN-LO in your code I changed the next line:

`videoDescription = "appsrc name=appsrc-video ! decodebin ! video/x-raw,width=1280,height=720,framerate=20/1  !  autovideosink";`

to 

`videoDescription = "appsrc name=appsrc-video ! decodebin ! video/x-raw  !  autovideosink";
`
and then it worked for me. This is the result:

![image](https://user-images.githubusercontent.com/36053387/176355991-c7d244fb-65ab-4161-8ea7-12dfd54da0f5.png)

![image](https://user-images.githubusercontent.com/36053387/176356021-76647d1a-b03c-4454-bb49-0c942424fa4c.png)
